internal/state/state.go

```diff
@@ -2,42 +2,65 @@ package state
 
 import (
 	"context"
-	"errors"
+	"fmt"
 	"io/ioutil"
 	"log"
 	"os"
 	"path/filepath"
+	"time"
 
+	"github.com/hetznercloud/cli/internal/hcapi2"
+	"github.com/hetznercloud/cli/internal/version"
 	"github.com/hetznercloud/hcloud-go/v2/hcloud"
 )
 
-type State struct {
+type State interface {
+	TokenEnsurer
+	ActionWaiter
+	context.Context
+	hcapi2.Client
+
+	WriteConfig() error
+	Config() *Config
+}
+
+type state struct {
+	context.Context
+	hcapi2.Client
+
 	Token         string
 	Endpoint      string
-	Context       context.Context
-	Config        *Config
 	ConfigPath    string
 	Debug         bool
 	DebugFilePath string
 
-	client *hcloud.Client
+	hcloudClient *hcloud.Client
+	config       *Config
 }
 
-func New() *State {
-	s := &State{
+func New() (State, error) {
+	s := &state{
 		Context:    context.Background(),
-		Config:     &Config{},
+		config:     &Config{},
 		ConfigPath: DefaultConfigPath,
 	}
 	if p := os.Getenv("HCLOUD_CONFIG"); p != "" {
 		s.ConfigPath = p
 	}
-	return s
+	if err := s.readConfig(); err != nil {
+		return nil, fmt.Errorf("unable to read config file %q: %s\n", s.ConfigPath, err)
+	}
+	s.readEnv()
+	s.hcloudClient = s.newClient()
+	s.Client = hcapi2.NewClient(s.hcloudClient)
```

Comment:
Great that we grouped the config loading together in the same place, felt a bit all over the place.

But I think we should refrain to have a huge New function, we should probably keep it split into multiple Public functions and run them in the main function.

This might help us with the poll interval flag, as we can delay the config loading and the client creation only after during the root command execution.

Reply:
Implemented in 6597bd5fe7c10b10bbbe860795530db30e8626b0